packages/engine/lib/error/src/lib.rs

```diff
@@ -0,0 +1,443 @@
+///     # Ok(())
+/// }
+/// ```
+#[must_use]
+pub struct Report<Context = ()> {
+    inner: Box<ReportImpl>,
+    _context: PhantomData<Context>,
+}
+
+/// A single error, contextual message, or error context inside of a [`Report`].
+///
+/// `Frame`s are an intrusive singly linked list. The head is pointing to the most recent error
+/// message or context, the tail is the root error created by [`Report::new()`],
+/// [`Report::from_context()`], or [`Report::from()`]. The list can be advanced by [`request`]ing
+/// [`tags::FrameSource`] or be iterated by calling [`Report::frames()`].
+///
+/// [`request`]: Self::request
+pub struct Frame {
+    error: ManuallyDrop<Box<FrameRepr>>,
+    location: &'static Location<'static>,
+    source: Option<Box<Frame>>,
+}
+
+/// `Result<T, Report<C>>`
+///
+/// A reasonable return type to use throughout an application if no scope is used.
+///
+/// The `Result` type can be used with one or two parameters.
+///
+/// # Examples
+///
+/// `Result` can also be used in `fn main()`:
+///
+/// ```
+/// # fn has_permission(_: usize, _: usize) -> bool { true }
+/// # fn get_user() -> Result<usize> { Ok(0) }
+/// # fn get_resource() -> Result<usize> { Ok(0) }
+/// use error::{ensure, Result};
+///
+/// fn main() -> Result<()> {
+///     let user = get_user()?;
+///     let resource = get_resource()?;
+///
+///     ensure!(
+///         has_permission(user, resource),
+///         "Permission denied for {user} accessing {resource}"
+///     );
+///
+///     //...
+///     # Ok(())
+/// }
+/// ```
+pub type Result<T, C = ()> = core::result::Result<T, Report<C>>;
+
+/// Extension trait for [`Result`][core::result::Result] to provide context information on
+/// [`Report`]s.
+pub trait ResultExt<T> {
+    /// Type of the resulting error `E` inside of [`Report<E>`][`Report`] when not providing an
+    /// error kind.
+    type Context;
+
+    /// Adds new context information to the [`Frame`] stack of a [`Report`].
+    fn wrap_err<M>(self, message: M) -> Result<T, Self::Context>
+    where
+        M: fmt::Display + fmt::Debug + Send + Sync + 'static;
+
+    /// Lazily adds new context information to the [`Frame`] stack of a [`Report`].
+    fn wrap_err_lazy<M, F>(self, message: F) -> Result<T, Self::Context>
+    where
+        M: fmt::Display + fmt::Debug + Send + Sync + 'static,
+        F: FnOnce() -> M;
+
+    /// Adds a new error kind to the [`Frame`] stack of a [`Report`].
+    fn provide_context<Context>(self, context: Context) -> Result<T, Context>
+    where
+        Context: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static;
+
+    /// Lazily adds a new error kind to the [`Frame`] stack of a [`Report`].
+    fn provide_context_lazy<E, Context>(self, context: Context) -> Result<T, E>
+    where
+        E: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
+        Context: FnOnce() -> E;
+}
```

Comment:
```suggestion
    /// Lazily adds a new error kind to the [`Frame`] stack of a [`Report`].
    fn provide_context_lazy<C, Context>(self, context: Context) -> Result<T, C>
    where
        E: Provider + fmt::Display + fmt::Debug + Send + Sync + 'static,
        Context: FnOnce() -> C;
}
```

lets try and be consistent with parameter args, might need a bit of a look through to see where you used to use ErrorKind and such instead of context